src/Raven.Server/Commercial/SetupManager.cs

```diff
@@ -0,0 +1,365 @@
+                                case 60:
+                                    progress.AddInfo("Stage2: Still Waiting... hang in there, just a few more seconds.");
+                                    break;
+                                case 120:
+                                    progress.AddInfo("Stage2: Something is wrong. You may wait a little longer but you might as well abort and try again.");
+                                    break;
+                                default:
+                                    progress.AddInfo("Stage2: Waiting...");
+                                    break;
+                            }
+                            onProgress(progress);
+                        }
+                        else if (registrationResult.Status == RegistrationStatus.Done)
+                            break;
+                        i++;
+                    }
+
+                    progress.Processed++;
+                    progress.AddInfo("Stage2: Finished.");
+                    progress.AddInfo($"Stage3: Completing Let's Encrypt challenge for domain {setupInfo.Domain}.");
+                    onProgress(progress);
+                    
+                    var challengeResult = await acmeClient.CompleteChallenge(challenge);
+
+                    for (i = 0; i < 15; i++)
+                    {
+                        authz = await acmeClient.GetAuthorization(challengeResult.Location);
+                        if (authz?.Data.Status != EntityStatus.Pending)
+                            break;
+
+                        await Task.Delay(250, token);
+                    }
+
+                    if (authz != null && authz.Data.Status != EntityStatus.Valid)
+                    {
+                        progress.AddError($"Stage3: Failed to authorize with Let\'s Encrypt: {authz.Data.Status}\r\n{authz.Json}"); 
+                        onProgress(progress);
+                        return null;
+                    }
+
+                    try
+                    {
+                        var csr = new CertificationRequestBuilder();
+                        csr.AddName($"CN={setupInfo.Domain}");
+                        csr.SubjectAlternativeNames.Add(setupInfo.Domain);
+                        
+                        // TODO make sure this is allowed, otherwise we need a challenge for every single node seperately
+                        foreach (var node in setupInfo.NodeSetupInfos)
+                        {
+                            csr.SubjectAlternativeNames.Add($"{node.NodeTag}.{setupInfo.Domain}");
+                        }
+                        var cert = await acmeClient.NewCertificate(csr);
+
+                        var pfxBuilder = cert.ToPfx();
+                        var certBytes = pfxBuilder.Build(setupInfo.Domain + " cert", "");
+                        var base64Cert = Convert.ToBase64String(certBytes);
+
+                        if (PlatformDetails.RunningOnPosix)
```

Comment:
I'm not very happy about this, because this may cause us to generate a lot of certificates while trying to setup on Linux, no?
I think we shouldn't do this validation here, but during the check phase later on.